sw/lib/sw/device/silicon_creator/rom_patch.c

```diff
@@ -0,0 +1,234 @@
+     * - It is valid, i.e. it's been fully programmed.
+     * - It has a strictly higher revision number.
+     */
+    if (rom_patch_lock_valid(patch_header) &&
+        (current_patch_revision > latest_patch_revision)) {
+      // Found a better patch, let's keep it.
+      latest_patch_addr = next_patch_addr;
+      latest_patch_revision = current_patch_revision;
+    }
+
+    next_patch_addr += current_patch_size_words * 4;
+  } while (current_patch_size_words > 0 && next_patch_addr < kRomPatchMaxAddr);
+
+  if (latest_patch_addr != kRomPatchInvalidAddr) {
+    patch_header = otp_read32(latest_patch_addr);
+    OT_DISCARD(rom_printf("Latest patch header 0x%x\n", patch_header));
+  }
+
+  return (uintptr_t)(latest_patch_addr);
+}
+
+static OT_WARN_UNUSED_RESULT rom_error_t rom_patch_verify_sig(
+    const rom_patch_t *patch, const hmac_digest_t *patch_digest) {
+  /*
+   * TODO: Verify that the loaded code matches the OTP signature.
+   */
+  return kErrorOk;
+}
+
+static OT_WARN_UNUSED_RESULT rom_error_t
+rom_patch_remap(const rom_patch_t *patch) {
+  for (uint32_t i = 0; i < RV_CORE_IBEX_PARAM_NUM_REGIONS; i++) {
+    const rom_patch_match_regs_t *match = &patch->table[i];
+    // If patch is not enabled, skip it.
+    if (!rom_patch_region_enabled(match)) {
+      continue;
+    }
+
+    uint32_t patch_size_bytes = (uint32_t)(rom_patch_region_size(match) << 2);
+    uint32_t m_base = rom_patch_region_m_base(match);
+    uint32_t r_base = rom_patch_region_r_base(match);
+
+    HARDENED_RETURN_IF_ERROR(
+        ibex_addr_remap_set(i, m_base, r_base, patch_size_bytes));
+    if (rom_patch_region_locked(match)) {
+      HARDENED_RETURN_IF_ERROR(ibex_addr_remap_lock(i));
+    }
+
+    OT_DISCARD(rom_printf(
+        "Configured Ibex remapping 0x%x->0x%x (0x%x bytes) Locked 0x%x\n",
+        m_base, r_base, patch_size_bytes, rom_patch_region_locked(match)));
+  }
+
+  return kErrorOk;
+}
+
+OT_WARN_UNUSED_RESULT rom_error_t
+rom_patch_apply(const uintptr_t patch_addr, hmac_digest_t *const patch_digest) {
+  size_t patch_code_offset = kRomPatchRegionPreambleSizeBytes;
+  uint32_t patch_preamble_bytes[kRomPatchRegionPreambleSizeBytes >> 2];
+  rom_patch_t *patch;
+
+  hmac_sha256_init();
+
+  // Read the patch preamble from OTP.
+  otp_read(patch_addr, patch_preamble_bytes,
+           kRomPatchRegionPreambleSizeBytes >> 2);
+  patch = (rom_patch_t *)(patch_preamble_bytes);
+
+  // The first header byte (LOCK_VALID & PROGRAM_START) is not signed.
+  hmac_sha256_update(patch_preamble_bytes + 1,
+                     kRomPatchRegionPreambleSizeBytes - 1);
+
+  // The patch code size must not be 0.
+  size_t patch_code_size_bytes = rom_patch_code_size_words(patch) << 2;
+  HARDENED_CHECK_NE(patch_code_size_bytes, 0);
+
+  // The remapping base address is the first entry in the match table.
+  const rom_patch_match_regs_t *match = &patch->table[0];
+  uint32_t remap_addr = rom_patch_region_r_base(match);
```

Comment:
I'm concerned that this code is copying the patch data to an address that is specified by the patch itself, and then validating the signature only after the copying has been performed. I think this has the potential to cause issues in the event of an OTP programming failure, and/or could perhaps be exploited in some way?